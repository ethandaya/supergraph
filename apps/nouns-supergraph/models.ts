import { z } from "zod";

export const NounSchema = z.object({
  id: z.string(),
  seed: z.nullable(z.number()),
  owner: z.string(),
});

export const BidSchema = z.object({
  id: z.string(),
  noun: z.string(),
  amount: z.string(),
  bidder: z.string(),
  auction: z.string(),
  txIndex: z.number(),
  blockNumber: z.number(),
  blockTimestamp: z.bigint(),
});

export const AuctionSchema = z.object({
  id: z.string(),
  noun: z.string(),
  amount: z.string(),
  startTime: z.bigint(),
  endTime: z.bigint(),
  settled: z.boolean(),
  bidder: z.string().nullable().default(null),
});

export const AccountSchema = z.object({
  id: z.string(),
  delegate: z.string(),
  tokenBalanceRaw: z.string(),
  tokenBalance: z.string(),
  totalTokensHeldRaw: z.string(),
  totalTokensHeld: z.string(),
});
