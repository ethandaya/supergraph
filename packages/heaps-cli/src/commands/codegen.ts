import { CAC } from "cac";
import { loadConfig } from "../utils/load";
import {
  EntityGenerator,
  EventGenerator,
  ModelGenerator,
} from "@heaps/generators";
import fs from "fs";
import { watch } from "chokidar";

export type CodegenOptions = {
  watch: boolean;
  pathToSchema: string;
  // TODO - design of store path needs a lot of work
  storeImportPath: string;
  pathToConfig: string;
  outputDir: string;
};
export function buildArtifacts(options: Omit<CodegenOptions, "watch">) {
  const config = loadConfig(options);
  const { pathToSchema, outputDir } = options;
  const modelGenerator = new ModelGenerator({
    schemaPath: pathToSchema,
    outputPath: outputDir + "/models.ts",
  });
  const entityGenerator = new EntityGenerator({
    isAsync: false,
    schemaPath: pathToSchema,
    outputPath: outputDir + "/schema.ts",
    storeImportPath: options.storeImportPath,
    modelImportPath: "./models",
  });
  for (const idx in config.sources) {
    const source = config.sources[idx];
    fs.mkdirSync(outputDir + `/${source.name}`, { recursive: true });
    const abi = JSON.parse(fs.readFileSync(source.abi, "utf-8"));
    const eventGenerator = new EventGenerator({
      abi,
      outputPath: outputDir + `/${source.name}/${source.name}.ts`,
    });
    eventGenerator.generate(true);
  }
  modelGenerator.generate(true);
  entityGenerator.generate(true);
}

export function codegen(options: CodegenOptions) {
  buildArtifacts(options);
  if (options.watch) {
    console.log("Watching for Changes...");
    watch(options.pathToSchema).on("change", async () => {
      console.log("Change detected, regenerating...");
      try {
        await codegen(options);
      } catch (e) {
        console.error("Error while regenerating");
        console.error(e);
      }
    });
  }
}

export function registerCodegenCommands(cli: CAC) {
  cli
    .command("codegen")
    .option("--watch", "Watch mode")
    .option("--pathToSchema", "Path to schema definition", {
      default: "./schema.graphql",
    })
    .option("--storeImportPath", "Path to store definition", {
      default: "../../store",
    })
    .option("--pathToConfig", "Path to supergraph config", {
      default: "./supergraph.json",
    })
    .option("--outputDir", "Directory to output artifacts too", {
      default: "./src/types",
    })
    .action(codegen);
  return cli;
}
