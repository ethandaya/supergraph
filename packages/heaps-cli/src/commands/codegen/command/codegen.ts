import { EntityGenerator, ModelGenerator } from "@heaps/generators";

type CodegenOptions = {
  watch: boolean;
  pathToSchema: string;
  outputDir: string;
};
export async function codegen(options: CodegenOptions) {
  const { pathToSchema, outputDir } = options;
  const modelGenerator = new ModelGenerator({
    schemaPath: pathToSchema,
    outputPath: outputDir + "/models.ts",
  });
  const entityGenerator = new EntityGenerator({
    isAsync: false,
    schemaPath: pathToSchema,
    outputPath: outputDir + "/schema.ts",
    storeImportPath: "./store",
    modelImportPath: outputDir + "./models",
  });
  await modelGenerator.generate(true);
  await entityGenerator.generate(true);
}
