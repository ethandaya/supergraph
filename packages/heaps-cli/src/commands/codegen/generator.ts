import {
  MethodDeclarationStructure,
  OptionalKind,
  Project,
  SourceFile,
} from "ts-morph";
import { z } from "zod";
import { makeGetterStatements, makeSetterStatements } from "./statements";

type GenerateOptions = {
  shouldSave: boolean;
};

type EntityGeneratorOptions = {
  models: { [key: string]: z.AnyZodObject };
  outputPath: string;
};

export class EntityGenerator {
  private project: Project;
  public targetFile: SourceFile;
  private readonly models: {
    [key: string]: z.AnyZodObject;
  } = {};

  constructor(options: EntityGeneratorOptions) {
    this.project = new Project();
    this.targetFile = this.project.createSourceFile(
      options.outputPath,
      undefined,
      {
        overwrite: true,
      }
    );
    this.models = options.models;
  }

  public generateImports() {
    this.targetFile.addImportDeclaration({
      namedImports: ["z"],
      moduleSpecifier: "zod",
    });
    this.targetFile.addImportDeclaration({
      namedImports: ["Entity", "KeyAccessError", "SQLiteStore as Store"],
      moduleSpecifier: "@heaps/engine",
    });
    this.targetFile.addImportDeclaration({
      namedImports: [...Object.keys(this.models)],
      moduleSpecifier: "../models",
    });
  }

  public generateTypeForSchema(name: string) {
    this.targetFile.addTypeAlias({
      name: name + "Model",
      type: `z.infer<typeof ${name}Schema>`,
    });
  }

  public generateClassForSchema(name: string, schema: z.ZodObject<any>) {
    const methods: OptionalKind<MethodDeclarationStructure>[] = [];

    methods.push({
      name: "load",
      isStatic: true,
      parameters: [
        {
          name: "id",
          type: "string",
        },
      ],
      returnType: `${name} | null`,
      statements: [
        `const data = store.get<${name}Model>("${name.toLowerCase()}", id);`,
        `if (!data) {`,
        `   return new ${name}(id);`,
        `}`,
        `return new ${name}(id, data);`,
      ],
    });

    for (const key in schema.shape) {
      const type = `${name}Model["${key}"]`;
      const isNullable = schema.shape[key].isNullable();
      methods.push({
        name: `get ${key}`,
        returnType: type,
        statements: makeGetterStatements({
          key,
          name,
          isNullable,
        }),
      });
      methods.push({
        name: `set ${key}`,
        parameters: [{ name: "value", type }],
        statements: makeSetterStatements(key),
      });
    }

    this.targetFile.addClass({
      name,
      isExported: true,
      extends: `Entity<${name}Model, typeof ${name}Schema>`,
      ctors: [
        {
          parameters: [
            {
              name: "id",
              type: "string",
            },
            {
              name: "data",
              type: `${name}Model`,
              hasQuestionToken: true,
            },
          ],
          statements: [
            `super(id, ${name}Schema, store)`,
            `this.data = { id, ...data } || {};`,
          ],
        },
      ],
      methods,
    });
  }

  public generateDefinitionsForModel(name: string, schema: z.ZodObject<any>) {
    this.generateTypeForSchema(name);
    this.generateClassForSchema(name, schema);
  }

  public generateGlobalDefinitions() {
    this.targetFile.addTypeAlias({
      name: "SchemaNames",
      type: Object.keys(this.models)
        .map((key) => `"${key.toLowerCase().replace("schema", "")}"`)
        .join(" | "),
    });
  }

  public generateStore() {
    this.targetFile.addStatements([
      `export const store = new Store<SchemaNames>("", {`,
      ...Object.keys(this.models).map(
        (key) => `  ${key.replace("Schema", "").toLowerCase()}: ${key},`
      ),
      `});`,
    ]);
  }

  public generate(options: GenerateOptions = { shouldSave: true }) {
    this.generateImports();
    this.generateGlobalDefinitions();
    this.generateStore();
    for (const name in this.models) {
      this.generateDefinitionsForModel(
        name.replace("Schema", ""),
        this.models[name]
      );
    }
    this.targetFile.prependWhitespace("\n\n");
    this.targetFile.insertText(
      0,
      "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY."
    );
    if (options.shouldSave) {
      this.save();
    }
  }

  public save() {
    this.targetFile.saveSync();
  }
}
