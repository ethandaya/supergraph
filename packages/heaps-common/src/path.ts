import path from "path";

export function getFullPath(filePath: string): string {
  return path.join(process.cwd(), filePath);
}

export function uncachedRequire(filePath: string): any {
  delete require.cache[require.resolve(filePath)];
  return require(filePath);
}
