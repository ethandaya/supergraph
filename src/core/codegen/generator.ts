import {
  MethodDeclarationStructure,
  OptionalKind,
  Project,
  SourceFile,
} from "ts-morph";
import { z } from "zod";
import { makeGetterStatements, makeSetterStatements } from "./statements";

export class EntityGenerator {
  private project: Project;
  public targetFile: SourceFile;
  private readonly models: z.ZodObject<any>[];
  constructor(private readonly pathToModels: string, outputPath: string) {
    this.project = new Project();
    this.targetFile = this.project.createSourceFile(outputPath, undefined, {
      overwrite: true,
    });
    this.models = require(this.pathToModels);
  }

  public generateImports() {
    this.targetFile.addImportDeclaration({
      namedImports: ["z"],
      moduleSpecifier: "zod",
    });
    this.targetFile.addImportDeclaration({
      namedImports: ["Entity", "KeyAccessError"],
      moduleSpecifier: "../../core/engine",
    });
    this.targetFile.addImportDeclaration({
      namedImports: [...Object.keys(this.models)],
      moduleSpecifier: "../models",
    });
    this.targetFile.addImportDeclaration({
      namedImports: ["SQLiteStore as Store"],
      moduleSpecifier: "../../core/store",
    });
    this.targetFile.addImportDeclaration({
      namedImports: ["formatModifiers"],
      moduleSpecifier: "../../core/format",
    });
  }

  public generateTypeForSchema(name: string) {
    this.targetFile.addTypeAlias({
      name: name + "Model",
      type: `z.infer<typeof ${name}Schema>`,
    });
  }

  public generateClassForSchema(name: string, schema: z.ZodObject<any>) {
    const methods: OptionalKind<MethodDeclarationStructure>[] = [];

    methods.push({
      name: "load",
      isStatic: true,
      parameters: [
        {
          name: "id",
          type: "string",
        },
      ],
      returnType: `${name} | null`,
      statements: [
        `const data = store.get<${name}Model>("${name.toLowerCase()}", id);`,
        `if (!data) {`,
        `   return new ${name}(id);`,
        `}`,
        `return new ${name}(id, data);`,
      ],
    });

    for (const key in schema.shape) {
      const type = `${name}Model["${key}"]`;
      const isNullable = schema.shape[key].isNullable();
      const isBoolean = schema.shape[key].isBoolean();
      methods.push({
        name: `get ${key}`,
        returnType: type,
        statements: makeGetterStatements({
          key,
          name,
          isNullable,
          isBoolean,
        }),
      });
      methods.push({
        name: `set ${key}`,
        parameters: [{ name: "value", type }],
        statements: makeSetterStatements(key),
      });
    }

    this.targetFile.addClass({
      name,
      isExported: true,
      extends: `Entity<${name}Model, typeof ${name}Schema>`,
      ctors: [
        {
          parameters: [
            {
              name: "id",
              type: "string",
            },
            {
              name: "data",
              type: `${name}Model`,
              hasQuestionToken: true,
            },
          ],
          statements: [
            `super(id, ${name}Schema, store)`,
            `this.data = { id, ...data } || {};`,
          ],
        },
      ],
      methods,
    });
  }

  public generateDefinitionsForModel(name: string, schema: z.ZodObject<any>) {
    this.generateTypeForSchema(name);
    this.generateClassForSchema(name, schema);
  }

  public generateGlobalDefinitions() {
    this.targetFile.addTypeAlias({
      name: "SchemaNames",
      type: Object.keys(this.models)
        .map((key) => `"${key.toLowerCase().replace("schema", "")}"`)
        .join(" | "),
    });
  }

  public generateStore() {
    this.targetFile.addStatements([
      `export const store = new Store<SchemaNames>("", {`,
      ...Object.keys(this.models).map(
        (key) => `  ${key.replace("Schema", "").toLowerCase()}: ${key},`
      ),
      `});`,
    ]);
    // this.targetFile.addVariableStatement({
    //   declarationKind: VariableDeclarationKind.Const,
    //   declarations: [
    //     {
    //       name: "store",
    //       initializer: `new Store<SchemaNames>("", schemaLookup)`,
    //       // .reduce(
    //       //   (acc, key) => ({
    //       //     ...acc,
    //       //     [key.replace("Schema", "").toLowerCase()]: key,
    //       //   }),
    //       //   {}
    //       // )
    //       // .toString()}})`,
    //     },
    //   ],
    // });
  }

  public generate() {
    this.generateImports();
    this.generateGlobalDefinitions();
    this.generateStore();
    for (const name in this.models) {
      this.generateDefinitionsForModel(
        name.replace("Schema", ""),
        this.models[name]
      );
    }
    this.targetFile.prependWhitespace("\n\n");
    this.targetFile.insertText(
      0,
      "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY."
    );
  }
}
