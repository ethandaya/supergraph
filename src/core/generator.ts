import { Project } from "ts-morph";

export function generateClassForSchema(schemaName: string, schema: any) {
  const project = new Project();

  // Create a new source file and add the seed schema to it
  const sourceFile = project.createSourceFile(
    "./src/mapping/generated/schema_.ts"
  );

  sourceFile.addTypeAlias({
    name: schemaName + "Model",
    type: `z.infer<typeof ${schemaName}Schema>`,
  });

  let methods = [];

  for (const key in schema.shape) {
    const type = `${schemaName}Model["${key}"]`;
    // const isNullable = schema.shape[key].isNullable;
    methods.push({
      name: `get ${key}`,
      returnType: type,
      statements: [
        `const value = this.get("${key}")`,
        `if (!value) {`,
        `  throw new KeyAccessError<${schemaName}>("${key}")`,
        `}`,
        `return value`,
      ],
    });
    methods.push({
      name: `set ${key}`,
      parameters: [
        {
          name: "value",
          type,
        },
      ],
      statements: `this.set("${key}", value);`,
    });
  }

  sourceFile.addClass({
    name: schemaName,
    isExported: true,
    extends: `Entity<${schemaName}Model>`,
    ctors: [
      {
        parameters: [
          {
            name: "id",
            type: "string | number",
          },
        ],
        statements: [`super(id, ${schemaName}Schema, store)`],
      },
    ],
    methods,
  });

  // await sourceFile.save();

  return sourceFile.getFullText();
}
