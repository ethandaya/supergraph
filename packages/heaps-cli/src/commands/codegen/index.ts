import { CAC } from "cac";
import { codegen } from "./command";

export function registerCodegenCommands(cli: CAC) {
  cli
    .command("codegen")
    .option("--watch", "Watch mode")
    .option("--pathToSchema", "Path to schema definition", {
      default: "./schema.graphql",
    })
    .option("--storeImportPath", "Path to store definition", {
      default: "../../store",
    })
    .option("--pathToConfig", "Path to supergraph config", {
      default: "./supergraph.json",
    })
    .option("--outputDir", "Directory to output artifacts too", {
      default: "./src/types",
    })
    .action(codegen);
  return cli;
}
