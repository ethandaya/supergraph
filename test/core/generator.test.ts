import { NounSchema } from "../../src/mapping/models";
import { EntityGenerator } from "../../src/core/codegen/generator";
import path from "path";
import { beforeEach, describe, expect, it } from "@jest/globals";

export function getFullPath(filePath: string): string {
  return path.join(process.cwd(), filePath);
}
describe("Generator", () => {
  let generator: EntityGenerator;

  beforeEach(() => {
    generator = new EntityGenerator(
      getFullPath("./src/mapping/models.ts"),
      getFullPath("./src/mapping/types/schema.ts")
    );
  });

  // it("should generate an entity & model definition for a schema", () => {
  //   generator.generateDefinitionsForModel("Seed", SeedSchema);
  //   expect(generator.targetFile.getFullText()).toMatchSnapshot();
  // });
  it("should generate an entity & model definition for a schema with a nullable field", () => {
    generator.generateDefinitionsForModel("Noun", NounSchema);
    expect(generator.targetFile.getFullText()).toMatchSnapshot();
  });
  // it("should generate an entity & model definition for a schema with nullables", () => {
  //   generator.generate();
  //   generator.targetFile.saveSync();
  //   // console.log(generator.targetFile.getFullText());
  // });
});
