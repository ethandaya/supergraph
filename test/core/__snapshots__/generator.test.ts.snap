// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator should generate an entity & model definition for a schema 1`] = `
"type SeedModel = z.infer<typeof SeedSchema>;

export class Seed extends Entity<SeedModel> {
    constructor(id: string | number) {
        super(id, SeedSchema, store)
    }

    get id(): SeedModel["id"] {
        const value = this.get("id")
        if (!value) {
          throw new KeyAccessError<Seed>("id")
        }

        return value
    }

    set id(value: SeedModel["id"]) {
        this.set("id", value);
    }

    get background(): SeedModel["background"] {
        const value = this.get("background")
        if (!value) {
          throw new KeyAccessError<Seed>("background")
        }

        return value
    }

    set background(value: SeedModel["background"]) {
        this.set("background", value);
    }

    get body(): SeedModel["body"] {
        const value = this.get("body")
        if (!value) {
          throw new KeyAccessError<Seed>("body")
        }

        return value
    }

    set body(value: SeedModel["body"]) {
        this.set("body", value);
    }

    get accessory(): SeedModel["accessory"] {
        const value = this.get("accessory")
        if (!value) {
          throw new KeyAccessError<Seed>("accessory")
        }

        return value
    }

    set accessory(value: SeedModel["accessory"]) {
        this.set("accessory", value);
    }

    get head(): SeedModel["head"] {
        const value = this.get("head")
        if (!value) {
          throw new KeyAccessError<Seed>("head")
        }

        return value
    }

    set head(value: SeedModel["head"]) {
        this.set("head", value);
    }

    get glasses(): SeedModel["glasses"] {
        const value = this.get("glasses")
        if (!value) {
          throw new KeyAccessError<Seed>("glasses")
        }

        return value
    }

    set glasses(value: SeedModel["glasses"]) {
        this.set("glasses", value);
    }
}
"
`;

exports[`Generator should get all classes in models.ts 1`] = `
"type SeedModel = z.infer<typeof SeedSchema>;

export class Seed extends Entity<SeedModel> {
    constructor(id: string) {
        super(id, SeedSchema, store)
    }

    get id(): SeedModel["id"] {
        const value = this.get("id")
        if (!value) {
          throw new KeyAccessError<Seed>("id")
        }

        return value
    }

    set id(value: SeedModel["id"]) {
        this.set("id", value);
    }

    get background(): SeedModel["background"] {
        const value = this.get("background")
        if (!value) {
          throw new KeyAccessError<Seed>("background")
        }

        return value
    }

    set background(value: SeedModel["background"]) {
        this.set("background", value);
    }

    get body(): SeedModel["body"] {
        const value = this.get("body")
        if (!value) {
          throw new KeyAccessError<Seed>("body")
        }

        return value
    }

    set body(value: SeedModel["body"]) {
        this.set("body", value);
    }

    get accessory(): SeedModel["accessory"] {
        const value = this.get("accessory")
        if (!value) {
          throw new KeyAccessError<Seed>("accessory")
        }

        return value
    }

    set accessory(value: SeedModel["accessory"]) {
        this.set("accessory", value);
    }

    get head(): SeedModel["head"] {
        const value = this.get("head")
        if (!value) {
          throw new KeyAccessError<Seed>("head")
        }

        return value
    }

    set head(value: SeedModel["head"]) {
        this.set("head", value);
    }

    get glasses(): SeedModel["glasses"] {
        const value = this.get("glasses")
        if (!value) {
          throw new KeyAccessError<Seed>("glasses")
        }

        return value
    }

    set glasses(value: SeedModel["glasses"]) {
        this.set("glasses", value);
    }
}
"
`;
